### PR TITLE
Tokotron: Tokenized TTS (lite version - minimal dependencies)

### PR DESCRIPTION
**Pull request by flexthink**
## What does this PR do?

Introduces a simple TTS architecture based on discrete speech representations from self-supervised models
Related to https://github.com/speechbrain/speechbrain/pull/2696

This version omits
* Evaluation (UTMOS/dWER)
* Utilities for the saving of samples and alignments

<details>
  <summary><b>Before submitting</b></summary>

- [x] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

File: speechbrain/lobes/models/discrete/Tokotron.py [hunk 1]
```diff
@@ -0,0 +1,2409 @@
+"""A simplistic Text-to-Speech model operating on
+discrete/tokenized audio representations, available in both
+Transformer and RNN flavours.
+
+NOTE: This model does not use the standa
+rd Transformer interface
+in order to make it usable as both as a full model and as a
+decoder-only model
```
Comment (pplantinga) [inline speechbrain/lobes/models/discrete/Tokotron.py:8]:
```suggestion
"""A Text-to-Speech model operating on discrete/tokenized audio
representations, available in both Transformer and RNN flavours.

NOTE: This model does not use the standard Transformer interface
in order to make it usable as both as a full model and as a decoder-only model
```

File: speechbrain/lobes/models/discrete/Tokotron.py [hunk 2]
```diff
@@ -0,0 +1,2409 @@
...
+        "alignments",
+    ],
+)
+
+TokotronDecoderOutput = namedtuple(
+    "TokotronDecoderOutput",
+    ["out", "gate_out", "dec_self_attn", "dec_attn", "alignments", "context"],
+)
+
+TokotronDecoderInfernceOutput = namedtuple(
```
Comment (pplantinga) [inline speechbrain/lobes/models/discrete/Tokotron.py:58]: Inference, not Infernce

File: speechbrain/lobes/models/discrete/Tokotron.py [hunk 3]
```diff
@@ -0,0 +1,2409 @@
...
+        "audio",
+        "length",
+        "dec_self_attn",
+        "dec_attn",
+        "alignments",
+        "p_eos",
+    ],
+)
+
+TokotronInfernceOutput = namedtuple(
```
Comment (pplantinga) [inline speechbrain/lobes/models/discrete/Tokotron.py:70]: Inference, not Infernce

File: speechbrain/lobes/models/discrete/Tokotron.py [hunk 4]
```diff
@@ -0,0 +1,2409 @@
...
+        if emb_size == out_size:
+            self.in_proj = nn.Identity()
+        else:
+            self.in_proj = Linear(
+                input_size=emb_size,
+                n_neurons=out_size,
+            )
+            nn.init.xavier_normal_(self.in_proj.w.weight)
+
+    """A simple embedding mechanism that adds the embedding to the inputs before the layer"""
```
Comment (pplantinga) [inline speechbrain/lobes/models/discrete/Tokotron.py:2142]:
```suggestion
```

File: recipes/LibriTTS/TTS/tokotron/train.py [hunk 5]
```diff
@@ -0,0 +1,989 @@
...
+
+        Returns
+        -------
+        loss : torch.Tensor
+            detached loss
+        """
+        loss = super().fit_batch(batch)
+        if self.hparams.lr_annealing_mode == "step":
+            self.hparams.lr_annealing(self.optimizer)
+        return loss
```
Comment (pplantinga) [inline recipes/LibriTTS/TTS/tokotron/train.py:419]: Maybe use this?
https://speechbrain.readthedocs.io/en/latest/API/speechbrain.core.html#speechbrain.core.Brain.on_fit_batch_end

File: recipes/LibriTTS/TTS/tokotron/train.py [hunk 6]
```diff
@@ -0,0 +1,989 @@
...
+        self,
+        epoch_counter,
+        train_set,
+        valid_set=None,
+        progressbar=None,
+        train_loader_kwargs={},
+        valid_loader_kwargs={},
+    ):
+        """Iterate epochs and datasets to improve objective.
+
```
Comment (pplantinga) [inline recipes/LibriTTS/TTS/tokotron/train.py:431]: Maybe instead of just copying the `brain` docstring this should state the changes that required overriding the default one.

**Review by pplantinga — COMMENTED** [review]
Thanks for your work minimizing the dependencies, this is still quite large however and will take some time to review, I only finished one pass and didn't have time to look at everything. I will also have to go and try to run the LibriTTS and LJSpeech recipes.

Overall, the code quality looks quite good, if a little verbose for my taste -- e.g. I'm not sure if the Additive Embedding and Null Embedding and Embedding Injection are really needed or if something simpler could be done. And some of the docstrings have extra spaces that don't quite match overall SpeechBrain docstring style.

Anything you can do to simplify and keep only the parts that are really necessary will be a huge help for my review, as well as future users!

**Comment by pplantinga** [comment]
Perhaps one thing we could do here is move the core changes to another PR: i.e. the four core (non-lobes) files in `nnet`, `utils`, and `loss`
